### PR TITLE
Fix/ns views not cleaned on removal

### LINKED ISSUE #1738: Potentiel leak: NativeScript views not cleaned up on removal
**Environment**
Provide version numbers for the following components (information can be retrieved by running `tns info` in your project folder or by inspecting the `package.json` of the project):
 - CLI: 5.2.0
 - Cross-platform modules:
 - Android Runtime: 5.2.0
 - iOS Runtime:5.2.0
 - Plugin(s):
 - NativeScript-Angular: ~7.2.1
 - Angular: ~7.2.0
 
**Describe the bug**

While working on https://github.com/NativeScript/nativescript-angular/pull/1728 I think I’ve found a couple of memory leaks in nativescript-angular.

***ListViewComponent/TemplatedItemsComponent items.***
In onItemLoading(args) https://github.com/NativeScript/nativescript-angular/blob/master/nativescript-angular/directives/templated-items-comp.ts#L141-L181 
~The `viewRef` is bound to the `args.view[NG_VIEW]`.~
~This is never cleared, so when the `ListViewComponent` is destroyed this reference still exists.
I suggest making this a `WeakRef`.~

The factory functions in TemplatedItemsComponent._templateMap creates a circular reference and should be cleared in `ngOnDestroy()`

***NativeScript views lives on after the angular app have been destroyed.***
The native views are destroyed as expected but according to the Chrome debugger the  NativeScript views lives on.
This is caused by the AppHostView not being cleared up on exit and the `firstChild`, `lastChild` and `nextSibling` references are not cleaned up.
The AppHostView is reused on next launch, so it shouldn’t itself be deleted but the children should.
Sidenote: AppHostView._ngAppRoot is a reference to the last Frame even after exit, this should be solved by deleting all its children on exit.

***ViewUtils.removeChild(parent, child) doesn’t remove children recursively.***
ViewUtils sets up the `firstChild`, `lastChild` and `nextSibling` references when a view is added, but doesn’t clear them up properly when a parent view is removes.

I suggest making removeFromQueue(parent, child) and removeFromVisualTree(parent, child)  recursive to clear up the references.

***PageRouterOutlet doesn’t clear up it’s children.***
`this.activated` should be destroyed in `ngOnDestroy()`

***NSLocationStrategy don’t have an ngOnDestroy()***
This leaves `correntOutlet` with a reference to the destroyed `PageRouterOutlet`.

**To Reproduce**

- Create a new nativescript angular app with: `tns create --ng NAME`
- Launch the app on android in debug mode and start the Chrome debugger.
- Close the app by tappen the back button, reopen and close the app a few times.
- Go to the memory tab in the Chrome debugger and take a heap snapshot.
- The snapshot will have N (= number of times the app has been opened) `Frame'/`ActionBar`/`Page`/`PageRouterOutlet`,  N*8 `StackLayout` etc. The native views will have been destroyed.
  - If you inspect the Views a little closer you should see they hold references to each other via `firstChild`, `lastChild` and `nextSibling`

**Note**: It doesn't seem to matter how many times I call `GC()` or how long I wait.

**Expected behavior**
- I expect the `GC()` to have collected the NativeScript Views.

**Sample project**

https://github.com/m-abs/tns-ng-app-lifecycle

**Additional context**

### PR DESCRIPTION
**Pull request by m-abs**
## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [x] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [x] You have signed the [CLA](http://www.nativescript.org/cla).
- [ ] All existing tests are passing: https://github.com/NativeScript/nativescript-angular/blob/master/DevelopmentWorkflow.md#running-the-tests
- [ ] Tests for the changes are included.

## What is the current behavior?

See the description in #1738 

## What is the new behavior?

- `ViewUtil.removeChild(parent, child)` now removes children recursively to clear references.
- `TemplatedItemsComponent.ngOnDestroy()` empties the `_templateMap` to remove a circular reference to the `TemplatedItemsComponent` in the factory functions.
- `PageRouterOutlet.ngOnDestroy()` will now destroy the activated componentRef.
- `NativeScriptRendererFactory.ngOnDestroy()` will now destroy the children of the rootNgView. 
- `NSLocationStrategy.ngOnDestroy()` removes its outlets.

Fixes #1738

Related to #1728

### CODE REVIEW

**Comment by etabakov** [comment]
@cla-bot check

**Comment by edusperoni** [comment]
Upon reviewing https://github.com/NativeScript/nativescript-angular/pull/1767, I realized that `view._modal` is never dealt with.

This was usually not an issue, but when we navigate out (and this actually happens with livesync too), the modal will never close. Maybe the modal should also be destroyed on https://github.com/NativeScript/nativescript-angular/blob/5b4f8ffdf25c2d9f9e62b5f0911c76bfe67faeaf/nativescript-angular/view-util.ts#L256

Or maybe this should be handled by tns-core-modules on `_removeView`:

https://github.com/NativeScript/NativeScript/blob/3c2c1d9b69eac347474241af9303232043125d37/tns-core-modules/ui/core/view-base/view-base.ts#L614

**Comment by m-abs** [comment]
I wanted to checkout the behavior, unfortunately I have trouble running `ng-sample`.

**Comment by m-abs** [comment]
Hi @edusperoni,

Do you have an example where `view.modal` isn't dealt with?

I have been trying to replicate the problem without any success.

**Comment by edusperoni** [comment]
I think this is mainly an issue when livesyncing. If you do:

`Page 1 -> Page 2 -> open Modal`

and livesync, it'll go back to page 1 and not close the modal. But I don't think it's something that should be considered in this PR

**Comment by VladimirAmiorkov** [comment]
test

**Comment by SvetoslavTsenov** [comment]
test

**Comment by SvetoslavTsenov** [comment]
test

**Comment by VladimirAmiorkov** [comment]
test

File: nativescript-angular/router/page-router-outlet.ts [hunk 1]
```diff
@@ -198,7 +198,16 @@ export class PageRouterOutlet implements OnDestroy { // tslint:disable-line:dire
             });
             this.locationStrategy.clearOutlet(this.frame);
         } else {
-            log("NSLocationStrategy.ngOnDestroy: no outlet available for page-router-outlet");
+            log("PageRouterOutlet.ngOnDestroy: no outlet available for page-router-outlet");
+        }
+
+        if (this.isActivated) {
+            const c = this.activated.instance;
+            this.activated.hostView.detach();
```
Comment (vakrilov) [inline nativescript-angular/router/page-router-outlet.ts]: Probably no need to call `detach` on component that will be destroyed anyway

**Comment by vakrilov** [comment]
Hey there @m-abs 
Looks like some really valid concerns are addressed in this PR.

The only change I didn't completely understand is [`ViewUtil.removeChild(parent, child) now removes children recursively to clear references`](https://github.com/NativeScript/nativescript-angular/pull/1740/files#diff-cfcd98afad2197d9ccdeb48144108797R175). My thinking is that if you want to remove a large portion of the UI tree you should be able to do by just removing the root of the subtree. Why is there a need to removing each child recursively? There might be performance implication if there are many children or deeper UI tree. In the same time

**Comment by m-abs** [comment]
@vakrilov 

> The only change I didn't completely understand is [`ViewUtil.removeChild(parent, child) now removes children recursively to clear references`](https://github.com/NativeScript/nativescript-angular/pull/1740/files#diff-cfcd98afad2197d9ccdeb48144108797R175).

That is unfortunately needed to clear up all references to `firstChild', `lastChild` and `nextSibling` from the UI tree below the view being removed. 
These references make it hard for the `GC` to clean om the disposed Views even after the native views have been destroyed.

If we have a UI tree like this:
```
View1
  View1_1
    View1_1_1
      View1_1_1_1
      View1_1_1_2
    View1_1.2
    View1_1_3
  View1_2
  View1_3
  View1_4
```

This gives us these references:
```
View1.firstChild = View1_1
View1.lastChild = View1_4
View1_1.nextSibling = View1_2
View1_2.nextSibling = View1_3
View1_3.nextSibling = View1_4
View1_1.firstChild = View1_1_1
View1_1.lastChild = View1_1.3
View1_1_1.firstChild = View1_1_1_1
View1_1_1.lastChild = View1_1_1_2
View1_1_1_1.nextSibling = View1_1_1_2
View1_1_1.nextSibling = View1_1_2
View1_1_3.nextSibling = View1_1_3
```

Removing `View1_1` will give us these:
```
View1.firstChild = View1_1
View1.lastChild = View1_4
View1_2.nextSibling = View1_3
View1_3.nextSibling = View1_4

View1_1_1.firstChild = View1_1_1_1
View1_1_1.lastChild = View1_1_1_2
View1_1_1_1.nextSibling = View1_1_1_2

View1_1_1.nextSibling = View1_1_2
View1_1_3.nextSibling = View1_1_3
```

**Comment by edusperoni** [comment]
Shouldn't only the parent and siblings be updated then?

```
View1.firstChild = View1_2 // checked and updated
View1.lastChild = View1_4 // checked and remained the same
View1_2.nextSibling = View1_3 // checked and remained the same
View1_3.nextSibling = View1_4 // checked and remained the same

View1_1_1.firstChild = View1_1_1_1
View1_1_1.lastChild = View1_1_1_2
View1_1_1_1.nextSibling = View1_1_1_2

View1_1_1.nextSibling = View1_1_2
View1_1_2.nextSibling = View1_1_3 // this would be updated if View1_1_3 were removed
```

Since there is no more reference to View1_1_1, the whole tree should be GCed.

**Comment by edusperoni** [comment]
I've added a simple example to illustrate this behavior: https://stackblitz.com/edit/typescript-pbquxj

How to check garbage collection:
1. Open Chrome on https://typescript-pbquxj.stackblitz.io
2. open dev tools and go into "Performance" (DO NOT USE STACKBLITZ'S CONSOLE)
3. Type in the console `document.weakSet`, all views will be there
4. after 5s a view will disappear
5. `document.weakSet` still shows all views
6. CLEAR THE CONSOLE (this is important, as chrome will keep the reference to the elements inside the WeakSet)
7. Press the Trash can (Collect garabage) in the performance tab a few times
8. check that the views vanished from `document.weakSet`

The same exercise could be done in the NativeScript runtime with `WeakRef`, but I believe it'll behave the same.

**Comment by vakrilov** [comment]
Indeed! The recursive remove does produce better results. I have also tried the same test with `*ngFor` instead of `ListView` and the results were consistently better.

**Note:** Adding `"markingMode": "none"` to the test project **decreased the GC time from ~1600ms to ~65ms, which is x25 fold improvement**. So I definitely recommend you to use it. In this setup the recursive update also performed slightly better.

That said - I'm sold on keeping the recursive remove. I have some other minor suggestions and we can merge this.

File: nativescript-angular/view-util.ts [hunk 2]
```diff
@@ -249,8 +264,11 @@ export class ViewUtil {
             this.removeLayoutChild(parent, child);
         } else if (isContentView(parent) && parent.content === child) {
             parent.content = null;
-            parent.lastChild = null;
-            parent.firstChild = null;
+        } else if (child.nodeName === "DetachedContainer") {
+            // Skip - DetachedContainer is... well detached from its parent
+            // Used with ListViews and other TemplatedItemsComponent views.
+        } else if (child.nodeName === "FormattedString" || child.nodeName === "Span") {
+            // Removing FormattedString and its Spans from a Label throws an exception
```
Comment (vakrilov) [inline nativescript-angular/view-util.ts]: Let's try to address these crashes in a separate issue/PR. I don't think it is a good practice to have element specific code inside the renderer.

File: nativescript-angular/view-util.ts [hunk 3]
```diff
@@ -249,8 +264,11 @@ export class ViewUtil {
             this.removeLayoutChild(parent, child);
         } else if (isContentView(parent) && parent.content === child) {
             parent.content = null;
-            parent.lastChild = null;
-            parent.firstChild = null;
+        } else if (child.nodeName === "DetachedContainer") {
```
Comment (vakrilov) [inline nativescript-angular/view-util.ts]: I would suggest more general approach. Let's use the `isDetachedElement` to restrict calling the [`removeFromVisualTree` at all](https://github.com/NativeScript/nativescript-angular/blob/2a64e87207c86bfa14533daaccea9d46ea007e1b/nativescript-angular/view-util.ts#L185) (this will work for all types of elements that are not added to the visual tree):
```
if (!isDetachedElement(extendedChild)) {
    this.removeFromVisualTree(extendedParent, extendedChild);
}
```

There is the same logic when [calling `addToVisualTree`](https://github.com/NativeScript/nativescript-angular/blob/2a64e87207c86bfa14533daaccea9d46ea007e1b/nativescript-angular/view-util.ts#L75-L78).

**Comment by m-abs** [comment]
> **Note:** Adding `"markingMode": "none"` to the test project **decreased the GC time from ~1600ms to ~65ms, which is x25 fold improvement**. So I definitely recommend you to use it. In this setup the recursive update also performed slightly better.

I tried `"markingMode": "none"` before and just now, it causes my test app to crash with: 

```
05-09 10:29:01.224 19269 19269 W System.err: com.tns.NativeScriptException: Attempt to use cleared object reference id=9224
05-09 10:29:01.224 19269 19269 W System.err:    at com.tns.Runtime.getJavaObjectByID(Runtime.java:1006)
05-09 10:29:01.224 19269 19269 W System.err:    at com.tns.Runtime.callJSMethodNative(Native Method)
05-09 10:29:01.224 19269 19269 W System.err:    at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1203)
05-09 10:29:01.224 19269 19269 W System.err:    at com.tns.Runtime.callJSMethodImpl(Runtime.java:1083)
05-09 10:29:01.224 19269 19269 W System.err:    at com.tns.Runtime.callJSMethod(Runtime.java:1070)
05-09 10:29:01.224 19269 19269 W System.err:    at com.tns.Runtime.callJSMethod(Runtime.java:1050)
05-09 10:29:01.224 19269 19269 W System.err:    at com.tns.Runtime.callJSMethod(Runtime.java:1042)
05-09 10:29:01.224 19269 19269 W System.err:    at com.tns.gen.android.widget.BaseAdapter_view_197_32_ListViewAdapter.getView(BaseAdapter_view_197_32_ListViewAdapter.java:40)
...
```
But that belongs in another issue :) 

> That said - I'm sold on keeping the recursive remove. I have some other minor suggestions and we can merge this.

Cool :)

File: nativescript-angular/view-util.ts (hunk 3 above)
Comment (m-abs) [reply nativescript-angular/view-util.ts]: Much better thank you.
I've made the suggested change.

**Comment by vakrilov** [comment]
Thanks @m-abs 
We will try to include this in a patch release next week along with NativeScript 5.4 release

**Comment by SvetoslavTsenov** [comment]
test ios@rc

**Comment by SvetoslavTsenov** [comment]
test ios@rc

**Comment by SvetoslavTsenov** [comment]
test ios#rc
